core/server/common/src/main/java/alluxio/master/audit/AsyncUserAccessAuditLogWriter.java

```diff
@@ -0,0 +1,123 @@
+ */
+
+package alluxio.master.audit;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.base.Preconditions;
+
+import java.util.concurrent.LinkedBlockingQueue;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * {@link AsyncUserAccessAuditLogWriter} writes user access audit log entries asynchronously.
+ */
+@ThreadSafe
+public final class AsyncUserAccessAuditLogWriter {
+  private static final int QUEUE_SIZE = 10000;
+  private static final Logger LOG =
+      LoggerFactory.getLogger(AsyncUserAccessAuditLogWriter.class);
+  private static final Logger AUDIT_LOG =
+      LoggerFactory.getLogger("AUDIT_LOG");
+  private volatile boolean mStopped;
+  /**
+   * A thread-safe linked-list-based queue with an optional capacity limit.
+   */
+  private LinkedBlockingQueue<AuditContext> mAuditLogEntries;
+
+  /**
+   * Background thread that performs actual log writing.
+   */
+  private Thread mLoggingWorkerThread;
+
+  /**
+   * Constructs an {@link AsyncUserAccessAuditLogWriter} instance.
+   */
+  public AsyncUserAccessAuditLogWriter() {
+    mAuditLogEntries = new LinkedBlockingQueue<>(QUEUE_SIZE);
+    mStopped = true;
+  }
+
+  /**
+   * Starts {@link AsyncUserAccessAuditLogWriter}.
+   */
+  public void start() {
+    synchronized (this) {
+      if (mStopped) {
+        Preconditions.checkState(mLoggingWorkerThread == null);
+        mStopped = false;
+        mLoggingWorkerThread = new Thread(new AuditLoggingWorker());
+        mLoggingWorkerThread.start();
+        LOG.info("AsyncUserAccessAuditLogWriter thread started.");
+      }
+    }
+  }
+
+  /**
+   * Stops {@link AsyncUserAccessAuditLogWriter}.
+   */
+  public void stop() {
+    synchronized (this) {
+      if (!mStopped) {
+        mLoggingWorkerThread.interrupt();
+        try {
+          mLoggingWorkerThread.join();
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+        } finally {
+          mStopped = true;
+          mLoggingWorkerThread = null;
+          LOG.info("AsyncUserAccessAuditLogWriter thread stopped.");
+        }
+      }
+    }
+  }
+
+  /**
+   * Appends an audit context.
+   *
+   * @param context the audit context to append
+   * @return true if append operation succeeds, false otherwise
+   */
+  public boolean append(AuditContext context) {
+    try {
+      mAuditLogEntries.put(context);
+    } catch (InterruptedException e) {
+      // Reset the interrupted flag and return because some other thread has
+      // told us not to wait any more.
+      Thread.currentThread().interrupt();
+      return false;
+    }
+    return true;
+  }
+
+  /**
+   * Consumer thread of the queue to perform actual logging of audit info.
+   */
+  private class AuditLoggingWorker implements Runnable {
+    public AuditLoggingWorker() {}
+
+    @Override
+    public void run() {
+      while (!mStopped) {
+        try {
+          AuditContext headContext = mAuditLogEntries.take();
```

Comment:
That requires us to find the performance bottleneck of existing operations and compare the overhead of the blocking queue. Will keep this in mind.